### DISCUSSION TITLE
Moving from mailing lists to discussions

### DISCUSSION THREAD

**Discussion by gtheler** [seamplex/feenox#2] [Announcements]
Hey all. Hopefully all questions, requirements and discussions will take place here, on GitHub discussions (pun).
The wasora mailing list at wasora@seamplex.com will be transitioned to this place and eventually closed, as this is a better way to stay up to date and to follow improvements and changes in the code.

Welcome to the FeenoX community!

**Comment by vitorvas** [comment]
Saludos hermano Germán and hello all,

Well, first of all, congrats. Like to see **feenox** coming into the light.

But let's go to the ugly part. As usual, even not planning to use feenox in the near future, I want to help testing it. And, for my surprise, I got a problem right in the ./configure step:

**configure.ac line 327**
`AC_CHECK_LIB([gsl],[gsl_blas_dgemm], [], AC_MSG_ERROR([GNU Scientific library libgsl-dev not found]))`


I decided to check by myself and got this [program](https://www.math.utah.edu/software/gsl/gsl-ref_200.html)

`#include <stdio.h>`
`#include <gsl/gsl_blas.h>`
` `
`int main (void)`
`{`
` double a[] = { 0.11, 0.12, 0.13, 0.21, 0.22, 0.23 };`
` double b[] = { 1011, 1012, 1021, 1022, 1031, 1032 };`
` double c[] = { 0.00, 0.00, 0.00, 0.00 };`
` `
` gsl_matrix_view A = gsl_matrix_view_array(a, 2, 3);`
` gsl_matrix_view B = gsl_matrix_view_array(b, 3, 2);`
` gsl_matrix_view C = gsl_matrix_view_array(c, 2, 2);`
` `
 ` /* Compute C = A B */`
` `
`  gsl_blas_dgemm (CblasNoTrans, CblasNoTrans,`
`                  1.0, &A.matrix, &B.matrix,`
`                  0.0, &C.matrix);`
` `
 ` printf("[ %g, %g\n", c[0], c[1]);`
 ` printf("  %g, %g ]\n", c[2], c[3]);`
` `
 ` return 0;  `
`}`

So, compiling and linking with gcc:

`gcc prog.c -o prog -lgsl`

and bang!

`[vitors@cochon tmp]$ gcc prog.c -o prog -lgsl
/bin/ld: /usr/lib/gcc/x86_64-redhat-linux/11/../../../../lib64/libgsl.so: undefined reference to `cblas_ctrmv'
(...)
/bin/ld: /usr/lib/gcc/x86_64-redhat-linux/11/../../../../lib64/libgsl.so: undefined reference to `cblas_dgemm'
/bin/ld: /usr/lib/gcc/x86_64-redhat-linux/11/../../../../lib64/libgsl.so: undefined reference to `cblas_srotmg'
/bin/ld: /usr/lib/gcc/x86_64-redhat-linux/11/../../../../lib64/libgsl.so: undefined reference to `cblas_dtrsm'
/bin/ld: /usr/lib/gcc/x86_64-redhat-linux/11/../../../../lib64/libgsl.so: undefined reference to `cblas_ccopy'
/bin/ld: /usr/lib/gcc/x86_64-redhat-linux/11/../../../../lib64/libgsl.so: undefined reference to `cblas_zsyrk'
/bin/ld: /usr/lib/gcc/x86_64-redhat-linux/11/../../../../lib64/libgsl.so: undefined reference to `cblas_cswap'
/bin/ld: /usr/lib/gcc/x86_64-redhat-linux/11/../../../../lib64/libgsl.so: undefined reference to `cblas_daxpy'
/bin/ld: /usr/lib/gcc/x86_64-redhat-linux/11/../../../../lib64/libgsl.so: undefined reference to `cblas_csyrk'
collect2: error: ld returned 1 exit status`

Linking error. But why, since I'm linking against gsl?

`[vitors@cochon tmp]$ gcc prog.c -o prog -lgsl -lcblas`
`[vitors@cochon tmp]$ ll prog`
`-rwxrwxr-x. 1 vitors vitors 24456 Sep  1 20:54 prog`
`[vitors@cochon tmp]$ `

It works now! What is the problem? I use Fedora (version 34 now) and it seems that when using gsl in Fedora the user also needs to link against cblas library. Yes, really weird. Feel free to follow this quite old [discussion](https://lists.fedoraproject.org/pipermail/devel/2015-February/208146.html).

So, I come with a (dirty) solution:

` AC_CHECK_LIB([gsl],[gsl_blas_dgemm], [], AC_MSG_ERROR([GNU Scientific library libgsl-dev not found]), [-lcblas])`
` gsl_version="from system"`
 `LIBS="$LIBS $LDFLAGS -lcblas"`

After changing these lines in `configure.ac` I got **feenox** working.

`[vitors@cochon feenox]$ ./feenox -V`
`FeenoX v0.1.34-g6b948b9-dirty `
`a free no-fee no-X uniX-like finite-element(ish) computational engineering tool`
` `
`Last commit date   : Tue Aug 31 20:07:36 2021 -0300`
`Build date         : Wed Sep 01 20:36:05 2021 -0300`
`Build architecture : linux-gnu x86_64`
`Compiler           : gcc (GCC) 11.2.1 20210728 (Red Hat 11.2.1-1)`
`Compiler flags     : -ggdb`
`Builder            : vitors@cochon`
`GSL version        : 2.6`
`SUNDIALS version   : 5.6.1`
`PETSc version      : Petsc Release Version 3.14.1, Nov 03, 2020` 
`PETSc arch         : arch-linux-c-debug`
`PETSc options      : --with-cc=gcc --with-cxx=g++ --with-fortran-bindings=0 --with-fc=0 --with-mpi=0`
`SLEPc version      : SLEPc Release Version 3.14.0, Sep 30, 2020`

You may ask: "Why not to open an issue and patch the fix?". Because it is not a fix. I'm just solving the problem for my Fedora box while ruining the `configure.ac` file for other distributions. The other obvious reason is that I expect the the gsl problem to get solved by the Fedora staff.

Of course, I'm open to suggestions. If someone knows how to make `autotools` check for the distribution and add want to live with many if statements to not break the current configure.ac file, I'm here to help. Meanwhile, I just want to share my (dirty) solution with the feenox community.

Regards and glad to share with you,

Vitor

**Comment by gtheler** [comment]
Hey Vitor! Long time no see!
Just a quick answer through the phone, more details tomorrow.
Actually it's not a GSL nor Fedora bug, it's a feature. GSL needs and
explicit link to gslcblas in case the user wants to use a faster/different
blas library (even a non-free one like MKL).

The thing is that PETSc already uses (and provides) a blas library which
can be used by GSL. So I tried to put a logic in configure.am such that
when PETSc is used, the blas provided by PETSc is used by GSL as well and
link to gslcblas if there's no PETSc. But I think I failed to do that (or
my trick does not work on Fedora). Will investigate tomorrow.
Feel free to open an issue so we can track the fix.

Thanks!

**Reply by vitorvas** [reply]
Well, now I feel bad for complaining on Fedora staff! I need to delete that, hahaha. It is just my lack of knowledge of the matter, as usual.
Anyway, I'll wait for you. If you think I should open an issue, I do it.

Regards!

**Reply by gtheler** [reply]
Moved the thread to another category: https://github.com/seamplex/feenox/discussions/4
